### ISSUE TITLE
How to convert appointment date from UTC to local timezone

### ISSUE THREAD

**Issue by mcisar** [jamesiarmes/php-ews#369]
This is probably incredibly simple and I'll be required to facepalm repeatedly but...

I have an inherited script that extracts calendar events from an exchange calendar using php-ews.  It has always displayed the start/end/reminder times for appointments in UTC (eg. 2016-12-05T19:00:00Z)  The calendar items do always come through with the proper timezone flag set (eg. (UTC-07:00) Mountain Time (US & Canada)).

However, I'd much rather stop thinking and just have the appointments displayed in the local timezone (in this case UTC-07:00).  

Is this something that should be handled by somehow telling the exchange server to send the results to me in UTC-07:00 rather than UTC...  or is this something I should be just converting on the script side before displaying the result?

Either way, prodding in the right direction and quick examples are appreciated, my PHP skills are a bit rusty (and the repeated banging of of my head on my desk has not helped this evening) GRIN.

Cheers & Thanks
Mike

**Comment by Garethp** [comment]
Have you tried sending a Timezone header with your request? In my fork (which I shamelessly plug, so I can help more people and provide a better quality library) [`garethp/php-ews`](https://github.com/Garethp/php-ews), it would be as simple as

```php
$api = API::withUsernameAndPassword('server', 'username', 'password', [
    'timezone' => 'Mountain Standard Time'
]);
```

I'm not entirely sure how to achieve that in this library, but I think you should be able to put the timezone somewhere in your request to get a calendar item.

I could also be wrong, and this doesn't affect returned DateTime, in that case you may want to be converting it once you've retrieved it by putting it in to a `DateTime` object, and then using the [`setTimezone`](http://php.net/manual/en/datetime.settimezone.php) method

**Comment by jamesiarmes** [comment]
@mcisar This library does not currently support setting the `TimeZoneContext` header, but this should be trivial to add, I will look into it this evening.

@Garethp Have you considered contributing some of your changes back to this library rather than constantly plugging your fork in my issue queue?

**Comment by Garethp** [comment]
@jamesiarmes I did actually, I [opened a PR about a year and a half ago](https://github.com/jamesiarmes/php-ews/pull/282), and had it open for about 10 months before closing it because the structure of our code was so different that I figured it would never be approved. I did also try emailing you and tweeting you, but I suspect you were too busy. In the mean time I've been trying to attend to both your issue list and my issue list and stack overflow.

I get it, you're incredibly busy and this library is undergoing massive changes, but whether you'll manage to get around to an issue or not is a bit of a gamble. So I try to help out, and suggest using my library because it's **a lot** easier for me to help people when they use my library

**Comment by jamesiarmes** [comment]
@Garethp My apologies then, I clearly missed your PR and additional contact. This was likely during a time when my ability to maintain this library had been limited. As I'm sure you've noticed, I have been working hard towards a 1.0.0 release, with the first beta release planned by the end of the year, so the structure isn't likely to change significantly at this point. Still, I'd be interested in what changes you have made that this library is currently missing (such as the `TimeZoneContext` header here). I'm also curious about any changes you may have made to make enumerations and array types a bit easier to work with.

**Comment by jamesiarmes** [comment]
@mcisar I have added a `setTimezone()` method to the client in 734f6a3ff7938dd9ad239f17e3cfa94dc832aa6c. This will allow you to set the timezone to be used for all requests.

**Comment by mcisar** [comment]
Thanks James, much appreciated.

Cheers,
Mike

**Comment by Garethp** [comment]
@jamesiarmes Sorry, you're right, I should keep the talk about my library out of your issues list. I've sent you an email, I'd love to collaborate and exchange ideas and techniques with you

**Comment by mcisar** [comment]
Ok I give :-)  I know my programming skills are as rusty as a sack of nails but went through and impressed myself by converting my script from the old version of php-ews that it was using to your current one and verified that my output was still what I expected.

Then I added this...
$mytimezone = 'Mountain Standard Time';

And changed this...
$client = new Client ($host, $username, $password, $version);
to this...
$client = new Client ($host, $username, $password, $version, $mytimezone);

Good news is nothing blows up, bad news is nothing changes from the output I was seeing before.  So I prayed before the grand Google and read several articles that I didn't understand... but all of their discussions of timezones seemed to make reference to Exchange 2007 SP1 (or higher) so making a brilliant connection I changed... 

$version = Client::VERSION_2007;
to
$version = Client::VERSION_2007_SP1;

Still no errors or anything thrown by the script but now zero output (so now it's not finding the appointments it found minutes before).  

I may be setting the timezone wrong, I may be setting it to the wrong thing

Umm... help :-)

Cheers,
Mike

**Comment by Garethp** [comment]
@mcisar The `$timezone` hasn't been added as a parameter on the constructor, but rather as a method once you've created it. Try:

```php
$client = new Client($host, $username, $password, $version);
$client->setTimezone($mytimezone);
```

**Comment by mcisar** [comment]
Thanks Gareth, gave that a try but still no joy.   Still in UTC with VERSION_2007 and no output with VERSION_2007_SP1

**Comment by Garethp** [comment]
Hm... can you try posting your full code here? You can format it by wrapping it in 

```
```php
//Code here
``
```

Blocks to make it more readable. With your whole code, maybe we can help you pinpoint your problem

**Comment by mcisar** [comment]
Here's the whole script except for the php start and end tags which the forum software seemed to have a hissyfit on.  Not sure why it's flipping in and out of code view either, but it looks like all of the code is there anyway.


``
require_once 'vendor/autoload.php';

use \jamesiarmes\PhpEws\Client;
use \jamesiarmes\PhpEws\Request\FindItemType;
use \jamesiarmes\PhpEws\Enumeration\ItemQueryTraversalType;
use \jamesiarmes\PhpEws\Type\ItemResponseShapeType;
use \jamesiarmes\PhpEws\Enumeration\DefaultShapeNamesType;
use \jamesiarmes\PhpEws\Type\CalendarViewType;
use \jamesiarmes\PhpEws\ArrayType\NonEmptyArrayOfBaseFolderIdsType;
use \jamesiarmes\PhpEws\Type\FolderIdType;
use \jamesiarmes\PhpEws\Enumeration\ResponseClassType;
use \jamesiarmes\PhpEws\Request\GetItemType;
use \jamesiarmes\PhpEws\Enumeration\BodyTypeResponseType;
use \jamesiarmes\PhpEws\ArrayType\NonEmptyArrayOfBaseItemIdsType;
use \jamesiarmes\PhpEws\Type\ItemIdType;

spl_autoload_register(
 function ($class) {
 $class = explode('_', $class);
 require_once 'php-ews/' .  $class[0] . '.php';
 }
 );

function getMessageBody($thisMID,$host,$username,$password){
// ***TODO*** Why does this only return the body sometimes with the new php-ews???
      //you need to make a whole new connection and separate request to get the Calendar body based off ItemId->Id:

          $client = new Client($host, $username, $password);
            //taken from our results above
               $message_id = $thisMID;
               $request = new GetItemType();

               $request->ItemShape = new ItemResponseShapeType();
               $request->ItemShape->BaseShape = DefaultShapeNamesType::ALL_PROPERTIES;
               $request->ItemShape->BodyType = BodyTypeResponseType::TEXT;

               $request->ItemIds = new NonEmptyArrayOfBaseItemIdsType();
               $request->ItemIds->ItemId = new ItemIdType();
               $request->ItemIds->ItemId->Id = $message_id;

               $response = $client->GetItem($request);
               //Return our message to the above loop:
               $message = $response->ResponseMessages->GetItemResponseMessage->Items->CalendarItem->Body->_;

               return $message;

}

 $today = Date("Y-m-d");
 $host = 'outlook.office365.com';
 $username = 'blah@blah.com';
 $password = 'supersecret';
 $version = '';
 $version = Client::VERSION_2007;
 $daysahead = '15';

$mytimezone = 'Mountain Standard Time';
//$localtimezone = 'America/Edmonton';

$client = new Client ($host, $username, $password, $version);
$client->setTimezone($mytimezone);

 $request = new FindItemType();

 $request->Traversal = ItemQueryTraversalType::SHALLOW;
 $request->ItemShape = new ItemResponseShapeType();
// $request->ItemShape->BaseShape = DefaultShapeNamesType::DEFAULT_PROPERTIES;
 $request->ItemShape->BaseShape = DefaultShapeNamesType::ALL_PROPERTIES;

 $request->CalendarView = new CalendarViewType();
 $request->CalendarView->StartDate = date ('c',strtotime("$today"));
 $request->CalendarView->EndDate = date ('c',strtotime("$today + $daysahead days"));

 $request->ParentFolderIds = new NonEmptyArrayOfBaseFolderIdsType();
// $request->ParentFolderIds->DistinguishedFolderId = new DistinguishedFolderIdType();
// $request->ParentFolderIds->DistinguishedFolderId->Id = DistinguishedFolderIdNameType::CALENDAR;

$folder_id = 'AZZZZG1jaXNhckBzdGFybWZZZZZubmV0AC4ZZZM+Tr+e6xBbTojH5kZZZZBsAQCOCg7VJpVZZZZ+u3+tYyMoZZZZZqYMbZZZZ';
$request->ParentFolderIds->FolderId = new FolderIdType();
$request->ParentFolderIds->FolderId->Id = $folder_id;

//$mailBox = new EmailAddressType();
//$mailBox->EmailAddress = "$EWScalendar";
//// $request->ParentFolderIds->DistinguishedFolderId->Mailbox = $mailBox;
//$request->ParentFolderIds->FolderId->Mailbox = $mailBox;

 $response = $client->FindItem($request);

// If only one item was found then we won't have an array from the response.

$response_messages = $response->ResponseMessages->FindItemResponseMessage;
if (!is_array($response_messages)) {
    $response_messages = array($response_messages);
}

// Iterate over the results, printing any error messages or contact ids.
foreach ($response_messages as $response_message) {
    // Make sure the request succeeded.
    if ($response_message->ResponseClass != ResponseClassType::SUCCESS) {
        $message = $response_message->ResponseCode;
// ***TODO*** why does this error message cause an error???
//        fwrite(STDERR, "Failed to search for events with \"$message\"\n");
        continue;
    }

//    $events = $response->ResponseMessages->FindItemResponseMessage->RootFolder->Items->CalendarItem;

    // If only one event was found, we won't have an array.
    $events = $response_message->RootFolder->Items->CalendarItem;
    if (!is_array($events)) {
        $events = array($events);
    }

    foreach ($events as $event){
        $id = $event->ItemId->Id;
        $change_key = $event->ItemId->ChangeKey;
        $start = $event->Start;
        $end = $event->End;
        $timezone = $event->TimeZone;
        $allday = $event->IsAllDayEvent;
        $duration = $event->Duration;
        $reminder = $event->ReminderDueBy;
        $reminderstart = $event->ReminderMinutesBeforeStart;
        $reminderset = $event->ReminderIsSet;
        $freebusy = $event->LegacyFreeBusyStatus;
        $categories = $event->Categories->String;
        $subject = $event->Subject;
        $location = $event->Location;
        $recurring = $event->IsRecurring;
        print("ID:  $id<br>");
        print("Start:  $start<br>");
        print("End:  $end<br>");
        print("Timezone:  $timezone<br>");
        print("All Day:  $allday<br>");
        print("Duration:  $duration<br>");
        print("Recurring:  $recurring<br>");
        print("Reminder:  $reminder<br>");
        print("Reminder Due:  $reminderstart<br>");
        print("Reminder Set:  $reminderset<br>");
        print("Free/Busy:  $freebusy<br>");
        print("Categories:  $categories<br>");
        print("Subject:  $subject<br>");
        print("Location:  $location<br>");
        //Call our Calendar body by using our custom function below called "getMessageBody":
        $calendarBody = getMessageBody($id,$host,$username,$password);
        print("Body:  $calendarBody<br>");
        print("<hr>");
    }

}
``

**Comment by jamesiarmes** [comment]
@mcisar To address two of your issues quick: You used two backticks instead of three to format your code, I've fixed it and it's now displaying correctly in GitHub. The `fwrite()` calls are failing because the example files are intended to be run from the command line and `fwrite()` doesn't work to write to stdout and stderr when run from a web server. As for your overall issue, I'm trying to reproduce now and I'll let you know what I find.

**Comment by jamesiarmes** [comment]
I have identified your issue, you are setting the version to Exchange 2007, however the `TimeZoneContext` header was not added until Exchange 2007 SP2. If your are indeed running SP2, you will need to update to the latest commit and use the new `VERSION_2009` constant. If you're using a version of exchange older than this, then you will need to use PHP to convert the values to the proper timezone. If you're using a newer version of exchange, then you should set the version to match.

A bit of advice on your code, I would recommend switching to PHP's object-oriented date functionality. In particular, you could clean up your date handling like so:

```php
 $today = new DateTime('today');
 
// intermediate code...

$request->CalendarView->StartDate = $today->format('c');
$request->CalendarView->EndDate = $today->add(new DateInterval('P1M'))->format('c');
```

**Comment by mcisar** [comment]
Thanks James, as for the backticks I just clicked on the "insert code" button in the editor.

As for the main issue, I'm on Office 365 so I'm assuming they will be using 2016, will try that.

Hrm ok now fixed the error handler also and am getting ErrorInvalidIdMalformedEwsLegacyIdFormat.. I'm going to guess I need to regenerate my folder ID for 2016 as well... that prog's on my other PC.    

Is there an easy way to enumerate a folder id from the actual text name programmatically (rather than just hacking in the actual folder ID as was done with the original script)?  In this case its a secondary calendar called "Bookings" underneath the main calendar folder.

Will also look into the date coding once I get everything else back to normal, I just did a barebones update of what it took to get the script running with the new library yesterday.
